core/modules/gtpu_parser.cc

```diff
@@ -0,0 +1,134 @@
+        set_gtp_parsing_attrs(&iph->src, &iph->dst, &tcph->src_port,
+                              &tcph->dst_port, (be32_t *)&_const_val,
+                              (be32_t *)&_const_val, &iph->protocol, p);
+        break;
+      case Ipv4::kUdp:
+        udph = (Udp *)((char *)iph + (iph->header_length << 2));
+        if (udph->dst_port == (be16_t)(UDP_PORT_GTPU)) {
+          gtph = (Gtpv1 *)(udph + 1);
+          // uint32_t teid = gtph->teid.value();
+          /* reuse iph, tcph, and udph for innser headers too */
+          iph = (Ipv4 *)((char *)gtph + gtph->header_length());
+          if (iph->protocol == Ipv4::kTcp) {
+            tcph = (Tcp *)((char *)iph + (iph->header_length << 2));
+            set_gtp_parsing_attrs(&iph->src, &iph->dst, &tcph->src_port,
+                                  &tcph->dst_port, (be32_t *)&_const_val,
+                                  (be32_t *)&_const_val, &iph->protocol, p);
+          } else if (iph->protocol == Ipv4::kUdp) {
+            udph = (Udp *)((char *)iph + (iph->header_length << 2));
+            set_gtp_parsing_attrs(&iph->src, &iph->dst, &udph->src_port,
+                                  &udph->dst_port, (be32_t *)&_const_val,
+                                  (be32_t *)&_const_val, &iph->protocol, p);
+          }
+        } else {
+          set_gtp_parsing_attrs(&iph->src, &iph->dst, &udph->src_port,
+                                &udph->dst_port, (be32_t *)&_const_val,
+                                (be32_t *)&_const_val, &iph->protocol, p);
+        }
+        break;
+      case Ipv4::kIcmp: {
+        set_gtp_parsing_attrs(&iph->src, &iph->dst, (be16_t *)&_const_val,
+                              (be16_t *)&_const_val, (be32_t *)&_const_val,
+                              (be32_t *)&_const_val, &iph->protocol, p);
+      } break;
+      default:
+        /* nothing here at the moment */
+        break;
+    }
+
+    EmitPacket(ctx, p, FORWARD_GATE);
+  }
+}
+/*----------------------------------------------------------------------------------*/
+CommandResponse GtpuParser::Init(const bess::pb::EmptyArg &) {
+  using AccessMode = bess::metadata::Attribute::AccessMode;
+  src_ip_id = AddMetadataAttr("src_ip", sizeof(uint32_t), AccessMode::kWrite);
+  dst_ip_id = AddMetadataAttr("dst_ip", sizeof(uint32_t), AccessMode::kWrite);
+  src_port_id =
+      AddMetadataAttr("src_port", sizeof(uint16_t), AccessMode::kWrite);
+  dst_port_id =
+      AddMetadataAttr("dst_port", sizeof(uint16_t), AccessMode::kWrite);
+  teid_id = AddMetadataAttr("teid", sizeof(uint32_t), AccessMode::kWrite);
+  tunnel_ip4_dst_id =
+      AddMetadataAttr("tunnel_ipv4_dst", sizeof(uint32_t), AccessMode::kWrite);
+  proto_id = AddMetadataAttr("ip_proto", sizeof(uint8_t), AccessMode::kWrite);
+
+  return CommandSuccess();
+}
+/*----------------------------------------------------------------------------------*/
+ADD_MODULE(GtpuParser, "gtpu_parser", "parsing module for gtp traffic")
```

Comment:
Should we rename this module and files as upf parser?

Reply:
GtpuParser makes more sense as we are dealing with gtp tunneled/untunneled traffic.